source/air/src/singular_manager.rs

```diff
@@ -22,29 +21,52 @@ impl SingularManager {
     pub fn launch(&self) -> SingularProcess {
         let mut child = std::process::Command::new(&self.singular_executable_name)
             .arg("--quiet")
+            // .arg("--no-tty") // these options are potentially useful for debugging purposes
+            // .arg("--no-shell")
             .stdin(std::process::Stdio::piped())
             .stdout(std::process::Stdio::piped())
             .spawn()
             .expect("could not execute Singular process");
-        let singular_pipe_stdout = BufReader::new(child.stdout.take().expect("take stdout"));
+
+        let child_stdout = child.stdout.take().expect("take stdout");
+        let singular_pipe_stdout = BufReader::new(child_stdout);
+
         let child_stdin = child.stdin.take().expect("take stdin");
         let (sender, receiver) = channel();
-        std::thread::spawn(move || writer_thread(receiver, child_stdin));
+        std::thread::spawn(move || singular_writer_thread(receiver, child_stdin));
         SingularProcess { requests: sender, singular_pipe_stdout }
     }
 }
 
+fn singular_writer_thread(requests: Receiver<Vec<u8>>, mut singular_pipe_stdin: ChildStdin) {
+    while let Ok(req) = requests.recv() {
+        singular_pipe_stdin
+            .write_all(&req)
+            .and_then(|_| writeln!(&singular_pipe_stdin))
+            .and_then(|_| singular_pipe_stdin.flush())
+            // The Singular process could die unexpectedly. In that case, we die too:
+            .expect("IO error: failure when sending data to Singular process across pipe");
```

Comment:
Can you elaborate on when you've seen this happen? Does it appear to be in response to a command we send?

Reply:
The old version was using `smt_process::writer_thread`, which sends an `echo` to Z3 after each command(s) for confirmation. Singular does not have `echo`, so sending that to a live Singular process will give an error. Further, Verus stops Singular by issuing `quit`. The process might not end immediately, so Verus might still send that `echo` to a process that is about to die, ending up with a broken pipe error message.